roachtest: add tpcc/mixed/nodes=3/headroom

tbg opened the pull request:
Add a mixed version cluster variant of the headroom test. This gives us
a better shot at catching problems that occur only in the mixed
configuration.

It already seems to be catching something because the import checks
immediately fail with

> Error: check failed: 3.3.2.2: pq: rpc error: code = Unimplemented desc
> = unknown service cockroach.sql.distsqlrun.DistSQL

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/36565)
<!-- Reviewable:end -->

tbg commented:
@jordanlewis can you (have someone) look into that DistSQL failure?

```
make bin/roachtest && ./bin/roachtest run --local tpcc/mixed/nodes=3/w=headroom
```

locally will repro it ~60s.

jordanlewis commented:
Yikes, thanks for catching this. I thought we already had mixed version cluster tests to protect us from stuff like this. I think the problem is that we moved the distsql protobuf definitions to a different package... ugh, I didn't realize that it would change the service name as well, though obviously that makes a lot of sense.

danhhz reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> good idea, thanks!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/36565#-:-LbhrtVV639izKnGb68I:b-1g9ts7)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @danhhz and @tbg)

___
*[pkg/cmd/roachtest/tpcc.go, line 118 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/36565#-Lbhqgth-ATVDN-_UX7h:-Lbhqgth-ATVDN-_UX7i:b-owv2sm) ([raw file](https://github.com/cockroachdb/cockroach/blob/109590d106c08fbe2d4efa1d86d08af118df5372/pkg/cmd/roachtest/tpcc.go#L118)):*
> ```Go
> 				// NB: binfetcher caches the downloaded files.
> 				binary, err := binfetcher.Download(ctx, binfetcher.Options{
> 					Binary:  "cockroach",
> ```

does running this roachtest overwrite the file named "cockroach" in the root of my checkout? if so, that's surprising
___
*[pkg/cmd/roachtest/tpcc.go, line 126 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/36565#-LbhqNTx1rnUDDPDEl48:-LbhqNTx1rnUDDPDEl49:b-8n84ez) ([raw file](https://github.com/cockroachdb/cockroach/blob/109590d106c08fbe2d4efa1d86d08af118df5372/pkg/cmd/roachtest/tpcc.go#L126)):*
> ```Go
> 					t.Fatalf("while fetching %s: %s", v, err)
> 				}
> 				c.Put(ctx, binary, "./cockroach", c.Node(i+1))
> ```

`roachprod stage <cluster> release <version>` will avoid downloading to your local machine, just to upload. I guess the --local case would still have to work like this
___
*[pkg/cmd/roachtest/tpcc.go, line 247 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/36565#-LbhqogpBOaDgYLbiiAU:-LbhqogpBOaDgYLbiiAV:bynxwi5) ([raw file](https://github.com/cockroachdb/cockroach/blob/109590d106c08fbe2d4efa1d86d08af118df5372/pkg/cmd/roachtest/tpcc.go#L247)):*
> ```Go
> 		// simulates a real production deployment in the middle of the migration
> 		// into a new cluster version.
> 		Name: "tpcc/mixed/nodes=3/w=headroom",
> ```

nit: our tendency for long roachtest names bugs me. I don't think nodes=3 is really essential to the name in the same way that mixed is. having the shorter cdc/one-thing ones has been working out for me. ymmv

if we do keep a description of the hardware in, i'd prefer the clusterSpec's String(), which has more info and is shorter (i should have done this with the other w=headroom)
___
*[pkg/cmd/roachtest/tpcc.go, line 251 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/36565#-LbhrBkN2s7thU2mkkWm:-LbhrBkO0bCPX56ST5wk:bp2v4a5) ([raw file](https://github.com/cockroachdb/cockroach/blob/109590d106c08fbe2d4efa1d86d08af118df5372/pkg/cmd/roachtest/tpcc.go#L251)):*
> ```Go
> 		// minimum on gce.
> 		MinVersion: maxVersion("v2.1.0", maybeMinVersionForFixturesImport(cloud)),
> 		Tags:       []string{`default`, `release_qualification`},
> ```

is release_qualification intentional? I suppose this shouldn't be any harder to harden than the other w=headroom once we fix the distsql bug you found


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
TFTRs, will update the PR later.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/36565#-:-LbhtC-400jM8FP6Z26y:bgzragm)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @danhhz and @tbg)

___
*[pkg/cmd/roachtest/tpcc.go, line 118 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/36565#-Lbhqgth-ATVDN-_UX7h:-Lbhs__dBr53TIklMYTg:b-t0ikco) ([raw file](https://github.com/cockroachdb/cockroach/blob/109590d106c08fbe2d4efa1d86d08af118df5372/pkg/cmd/roachtest/tpcc.go#L118)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

does running this roachtest overwrite the file named "cockroach" in the root of my checkout? if so, that's surprising
</blockquote></details>

Nope, this downloads into your temp dir.
___
*[pkg/cmd/roachtest/tpcc.go, line 126 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/36565#-LbhqNTx1rnUDDPDEl48:-LbhsjXG2VnTtmkYe8ss:b-vo1clt) ([raw file](https://github.com/cockroachdb/cockroach/blob/109590d106c08fbe2d4efa1d86d08af118df5372/pkg/cmd/roachtest/tpcc.go#L126)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

`roachprod stage <cluster> release <version>` will avoid downloading to your local machine, just to upload. I guess the --local case would still have to work like this
</blockquote></details>

Hmm good point. I'll leave as is though until `roachprod stage local release "xxx"` works (I think you're saying that it won't work today, at least not for non-linux users)
___
*[pkg/cmd/roachtest/tpcc.go, line 251 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/36565#-LbhrBkN2s7thU2mkkWm:-LbhsyAD8dwXn-qcpP9h:bqlx8bw) ([raw file](https://github.com/cockroachdb/cockroach/blob/109590d106c08fbe2d4efa1d86d08af118df5372/pkg/cmd/roachtest/tpcc.go#L251)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

is release_qualification intentional? I suppose this shouldn't be any harder to harden than the other w=headroom once we fix the distsql bug you found
</blockquote></details>

Not intentional with this PR, though I think generally the next two tests we want to include is one that runs with more than three nodes (to catch rebalancing problems) and one that runs mixed on more than three nodes.


<!-- Sent from Reviewable.io -->

tbg committed:
roachtest: add tpcc/mixed-headroom on five nodes

Add a mixed version cluster variant of the headroom test. This gives us
a better shot at catching problems that occur only in the mixed
configuration.

It already caught a DistSQL incompatibility (since fixed):

> Error: check failed: 3.3.2.2: pq: rpc error: code = Unimplemented desc
> = unknown service cockroach.sql.distsqlrun.DistSQL

Release note: None

tbg reviewed (commented):
Running this for good measure, so far it's happily importing. Will merge this tomorrow if all goes well (and if there aren't any more comments to address).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/36565#-:-Lbio7Zo3nccd9z320ig:b-ryu905)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @danhhz)

___
*[pkg/cmd/roachtest/tpcc.go, line 247 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/36565#-LbhqogpBOaDgYLbiiAU:-LbigKxaDbEv5yNMV3cn:b-ul8u2a) ([raw file](https://github.com/cockroachdb/cockroach/blob/109590d106c08fbe2d4efa1d86d08af118df5372/pkg/cmd/roachtest/tpcc.go#L247)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

nit: our tendency for long roachtest names bugs me. I don't think nodes=3 is really essential to the name in the same way that mixed is. having the shorter cdc/one-thing ones has been working out for me. ymmv

if we do keep a description of the hardware in, i'd prefer the clusterSpec's String(), which has more info and is shorter (i should have done this with the other w=headroom)
</blockquote></details>

Let's try this out with both of these tests (done?) but I don't want to go on a rename rampage in this PR.
___
*[pkg/cmd/roachtest/tpcc.go, line 251 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/36565#-LbhrBkN2s7thU2mkkWm:-LbigjV90EpHSbJx5ojm:b5dx3r0) ([raw file](https://github.com/cockroachdb/cockroach/blob/109590d106c08fbe2d4efa1d86d08af118df5372/pkg/cmd/roachtest/tpcc.go#L251)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Not intentional with this PR, though I think generally the next two tests we want to include is one that runs with more than three nodes (to catch rebalancing problems) and one that runs mixed on more than three nodes.
</blockquote></details>

I removed the label, but I added an additional node to have some rebalancing during the test.


<!-- Sent from Reviewable.io -->

danhhz reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/36565)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @danhhz and @tbg)

___
*[pkg/cmd/roachtest/tpcc.go, line 247 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/36565#-LbhqogpBOaDgYLbiiAU:-LbjTYqZ7TbrdRanOiux:bsh6e5y) ([raw file](https://github.com/cockroachdb/cockroach/blob/109590d106c08fbe2d4efa1d86d08af118df5372/pkg/cmd/roachtest/tpcc.go#L247)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Let's try this out with both of these tests (done?) but I don't want to go on a rename rampage in this PR.
</blockquote></details>

agreed. renaming breaks the open github issue matching, so needs to be done carefully
___
*[pkg/cmd/roachtest/tpcc.go, line 251 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/36565#-LbhrBkN2s7thU2mkkWm:-LbjTz7PAG7XOr7Vew02:bn1wz0t) ([raw file](https://github.com/cockroachdb/cockroach/blob/109590d106c08fbe2d4efa1d86d08af118df5372/pkg/cmd/roachtest/tpcc.go#L251)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

I removed the label, but I added an additional node to have some rebalancing during the test.
</blockquote></details>

:thumbsup:


<!-- Sent from Reviewable.io -->

tbg commented:
The test passed. TFTRs!

bors r=danhhz
